### PR TITLE
Handle Dayforce feed variants and add 27 employers

### PR DESCRIPTION
**Pull request by kalil0321**
## Summary
- collapse equivalent multi-location and multilingual Dayforce feed rows by requisition instead of rejecting an otherwise valid board
- still fail closed when one culture reuses a requisition ID for conflicting titles or employers
- prefer English variants, preserve all exposed cultures and locations in `raw`, and recognize both ISO alpha-2 and alpha-3 country codes
- add 27 validated direct-employer Dayforce boards
- compare duplicate employers from the original feed before applying the canonical CSV name
- retain an available description when a preferred-language variant is empty

## Duplicate controls
- excluded five ACV subsidiary boards because 18/24 jobs overlapped the main ACV board
- excluded two secondary Alliant boards because their jobs overlapped the broader `Alliant` board
- excluded Aline, Adare Pharma Solutions, and AGCO because those employers already exist under BambooHR, Workable, and SuccessFactors
- did not add five empty boards or six boards that failed live API validation

## Live validation (2026-07-27)
- 27/27 added boards completed the full scraper path
- 3,232 jobs, 3,232 unique ATS IDs, and 3,232 unique URLs
- 3,232/3,232 descriptions populated
- 3,218/3,232 locations populated
- regional split: 3,184 North America, 28 Europe, 6 Asia, 14 unknown

## Tests
- `PYTHONPATH=src uv run --extra test --with-requirements pipeline/requirements.txt pytest -q tests/test_dayforce.py tests/test_dayforce_contracts.py tests/test_resolve.py`
- 110 passed
- `PYTHONPATH=src uv run --extra dev ruff check src/ats_scrapers/scrapers/dayforce.py tests/test_dayforce.py`
- all checks passed

---
## Summary by cubic
Collapses multilingual and multi-location Dayforce feed rows into a single job per requisition, prioritizes populated descriptions before English, and adds 27 validated direct-employer boards to expand coverage.

- **New Features**
  - Collapse duplicate Dayforce rows by `ReferenceNumber`; return one job per requisition.
  - Select a variant with a description first; then prefer English. Preserve cultures in `raw.AvailableCultures` and locations in `raw.AllLocations`.
  - Accept ISO‑2 and ISO‑3 country codes for ISO/region normalization.
  - Add 27 validated direct-employer boards; exclude overlapping subsidiaries and invalid/empty boards.

- **Bug Fixes**
  - Preserve feed employer when checking duplicates; fail closed on conflicting titles or employers (even with a seed override).
  - Avoid dropping the only populated description when the English row is empty.
  - Add tests covering description-first selection, English preference, multi-location preservation, and duplicate-employer failure.

<sup>Written for commit 5ee3643b41ec6bb211d4ee5bf3d16e1881dee6c2. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/kalil0321/ats-scrapers/pull/231?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

<h3>Greptile Summary</h3>

This PR expands Dayforce coverage and changes duplicate requisition handling.
- Adds 27 Dayforce employer boards to the canonical tenant list.
- Collapses multilingual and multi-location feed rows by requisition.
- Prefers English variants and preserves available cultures and locations in raw data.
- Accepts ISO alpha-2 and alpha-3 country codes.
- Adds tests for equivalent, multilingual, and multi-location variants.

<h3>Confidence Score: 4/5</h3>

The description-loss path should be fixed before merging because valid multilingual requisitions can be published without their available job content.

English culture is ranked before description availability, and selection retains only one variant’s fields, so a populated non-English description is discarded when the preferred English row is empty.

**Files Needing Attention:** src/ats_scrapers/scrapers/dayforce.py

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/ats_scrapers/scrapers/dayforce.py | Adds variant grouping, conflict detection, selection ranking, aggregate raw metadata, and dual ISO-code normalization; ranking can discard the only populated description. |
| tests/test_dayforce.py | Covers duplicate collapsing, English selection, and location preservation, but not an empty preferred-language description. |
| ats-companies/dayforce.csv | Adds 27 canonical Dayforce tenant entries without an identified code-level defect. |

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
### Issue 1
src/ats_scrapers/scrapers/dayforce.py:273-278
**Ranking discards available descriptions**

When an English variant has no description and a non-English variant has one, the culture-first ranking selects the empty English row and retains only that row’s fields, causing the published job to omit a description that Dayforce returned.

---

For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
`````

</details>

<sub>Reviews (1): Last reviewed commit: ["Handle Dayforce feed variants and add va..."](https://github.com/kalil0321/ats-scrapers/commit/53cd6661d0c0889dc42f99f571a894aae1c0e001) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=47724510)</sub>

> Greptile also left **1 inline comment** on this PR.

**Context used:**

- Knowledge Base — [Scraper Adapters](https://app.greptile.com/kalil0321/-/custom-context/knowledge-base/kalil0321/ats-scrapers/-/docs/scraper-adapters.md)
- Knowledge Base — [Run Pipeline: Tenant List → Flat Jobs CSV](https://app.greptile.com/kalil0321/-/custom-context/knowledge-base/kalil0321/ats-scrapers/-/docs/run-pipeline-script.md)

### CODE REVIEW

File: src/ats_scrapers/scrapers/dayforce.py [hunk 1]
```diff
@@ -217,6 +222,62 @@ def _parse_job(self, item: dict[str, Any]) -> Job:
...
+    variants: list[tuple[Job, dict[str, Any]]],
+) -> Job:
+    cultures_to_titles: dict[str, set[str]] = {}
+    companies: set[str] = set()
+    for job, item in variants:
+        culture = (_string(item.get("CultureCode")) or "").casefold()
+        cultures_to_titles.setdefault(culture, set()).add(
+            " ".join(job.title.split()).casefold()
+        )
+        companies.add(" ".join(job.company.split()).casefold())
```
[the remark this answers is not in the document]
Comment (kalil0321) [reply src/ats_scrapers/scrapers/dayforce.py]: Fixed: duplicate conflict detection now compares the original feed employer fields before applying the canonical seed name. Added a regression test.

File: src/ats_scrapers/scrapers/dayforce.py [hunk 2]
```diff
@@ -217,6 +222,62 @@ def _parse_job(self, item: dict[str, Any]) -> Job:
...
+
+def _variant_rank(job: Job, item: dict[str, Any]) -> tuple[bool, bool, bool, bool]:
+    culture = (_string(item.get("CultureCode")) or "").casefold()
+    country = (_string(item.get("Country")) or "").upper()
+    return (
+        culture == "en-us",
+        culture.startswith("en-"),
+        country in _COUNTRY_METADATA,
+        bool(job.description),
+    )
```
[the remark this answers is not in the document]
Comment (kalil0321) [reply src/ats_scrapers/scrapers/dayforce.py:283]: Fixed: variants with descriptions now outrank empty variants; English remains preferred among equally complete variants. Added a regression test.
